File: services/research/neurogrip/windowing.py
"""Framing a continuous multi-channel stream into overlapping analysis windows.

Mirrored in packages/core/src/windowing.ts. The 20 ms hop (rather than the
100 ms in the project report) is what makes progressive actuation possible:
at a 10 Hz decision rate there is no evidence to accumulate between decisions.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class WindowConfig:
    window_ms: int = 200
    hop_ms: int = 20
    sampling_rate_hz: int = 2000

    @property
    def window_samples(self) -> int:
        if self.window_ms <= 0:
            raise ValueError("window_ms must be positive")
        return int(self.window_ms * self.sampling_rate_hz / 1000)

    @property
    def hop_samples(self) -> int:
        if self.hop_ms <= 0:
            raise ValueError("hop_ms must be positive")
        return int(self.hop_ms * self.sampling_rate_hz / 1000)


def window_start_indices(n_samples: int, config: WindowConfig) -> list[int]:
    """Start offsets of every complete window. A partial tail is dropped."""
    width = config.window_samples
    hop = config.hop_samples
    if n_samples < width:
        return []
    return list(range(0, n_samples - width + 1, hop))


def frame_windows(channels: np.ndarray, config: WindowConfig) -> np.ndarray:
    """Frame (n_channels, n_samples) into (n_windows, n_channels, window_samples).

    A trailing partial window is dropped rather than zero-padded: a zero-filled
    tail yields features no real contraction could produce and would pollute
    both training data and the drift statistics.
    """
    signal = np.asarray(channels, dtype=np.float32)
    if signal.ndim != 2:
        raise ValueError("EMG input must have shape (n_channels, n_samples)")

    n_channels, n_samples = signal.shape
    width = config.window_samples
    starts = window_start_indices(n_samples, config)
    if not starts:
        return np.zeros((0, n_channels, width), dtype=np.float32)

    return np.stack([signal[:, start : start + width] for start in starts])

File: packages/core/src/windowing.ts
/**
 * Streaming front end: window geometry and a preallocated ring buffer.
 *
 * Mirrors services/research/neurogrip/windowing.py for geometry. The 20 ms hop
 * (rather than the report's 100 ms) is what makes progressive actuation
 * possible: at a 10 Hz decision rate there is nothing to accumulate between
 * decisions.
 */

export interface WindowConfig {
  windowMs: number;
  hopMs: number;
  samplingRateHz: number;
}

export const DEFAULT_WINDOW_CONFIG: WindowConfig = {
  windowMs: 200,
  hopMs: 20,
  samplingRateHz: 2000,
};

export function windowSamples(config: WindowConfig): number {
  if (config.windowMs <= 0) throw new RangeError('windowMs must be positive');
  return Math.trunc((config.windowMs * config.samplingRateHz) / 1000);
}

export function hopSamples(config: WindowConfig): number {
  if (config.hopMs <= 0) throw new RangeError('hopMs must be positive');
  return Math.trunc((config.hopMs * config.samplingRateHz) / 1000);
}

/**
 * Fixed-capacity multi-channel ring buffer.
 *
 * Preallocates and never grows. Allocating inside the 50 Hz inference loop
 * invites garbage-collection pauses, which would land directly in the P95
 * latency figure this project treats as a safety property.
 */
export class MultiChannelRingBuffer {
  private readonly buffers: Float64Array[];
  private readonly scratch: Float64Array[];
  private writeIndex = 0;
  private filled = 0;

  constructor(
    public readonly nChannels: number,
    public readonly capacitySamples: number,
  ) {
    if (nChannels < 1) throw new RangeError('nChannels must be at least 1');
    if (capacitySamples < 1) throw new RangeError('capacitySamples must be at least 1');
    this.buffers = Array.from({ length: nChannels }, () => new Float64Array(capacitySamples));
    this.scratch = Array.from({ length: nChannels }, () => new Float64Array(capacitySamples));
  }

  get available(): number {
    return this.filled;
  }

  push(frame: Float64Array[]): void {
    if (frame.length !== this.nChannels) {
      throw new RangeError(`expected ${this.nChannels} channels, received ${frame.length}`);
    }
    const count = frame[0]!.length;
    for (const channel of frame) {
      if (channel.length !== count) throw new RangeError('channels must be equal length');
    }

    for (let i = 0; i < count; i++) {
      const slot = (this.writeIndex + i) % this.capacitySamples;
      for (let c = 0; c < this.nChannels; c++) {
        this.buffers[c]![slot] = frame[c]![i]!;
      }
    }
    this.writeIndex = (this.writeIndex + count) % this.capacitySamples;
    this.filled = Math.min(this.capacitySamples, this.filled + count);
  }

  /**
   * The most recent `width` samples per channel, oldest first, or null if not
   * enough data has arrived. The returned arrays are reused scratch views and
   * are invalidated by the next call - copy them if you need to retain them.
   */
  readLatestWindow(width: number): Float64Array[] | null {
    if (width < 1) throw new RangeError('width must be positive');
    if (width > this.capacitySamples) {
      throw new RangeError('width exceeds buffer capacity');
    }
    if (this.filled < width) return null;

    const start = (this.writeIndex - width + this.capacitySamples) % this.capacitySamples;
    const out: Float64Array[] = [];
    for (let c = 0; c < this.nChannels; c++) {
      const view = this.scratch[c]!.subarray(0, width);
      for (let i = 0; i < width; i++) {
        view[i] = this.buffers[c]![(start + i) % this.capacitySamples]!;
      }
      out.push(view);
    }
    return out;
  }
}
